vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+#include "WinRTWebSocketResource.h"
+
+#include <winrt/Windows.Security.Cryptography.h>
+#include <Unicode.h>
+#include <Utilities.h>
+
+// Standard Library
+#include <future>
+
+using Microsoft::Common::Unicode::Utf8ToUtf16;
+using Microsoft::Common::Unicode::Utf16ToUtf8;
+using Microsoft::Common::Utilities::CheckedReinterpretCast;
+
+using std::function;
+using std::size_t;
+using std::string;
+using std::vector;
+
+using winrt::fire_and_forget;
+using winrt::hresult;
+using winrt::hresult_error;
+using winrt::resume_background;
+using winrt::Windows::Foundation::Uri;
+using winrt::Windows::Networking::Sockets::IWebSocket;
+using winrt::Windows::Networking::Sockets::MessageWebSocket;
+using winrt::Windows::Networking::Sockets::MessageWebSocketMessageReceivedEventArgs;
+using winrt::Windows::Networking::Sockets::SocketMessageType;
+using winrt::Windows::Networking::Sockets::WebSocketClosedEventArgs;
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+using winrt::Windows::Storage::Streams::DataReader;
+using winrt::Windows::Storage::Streams::DataWriter;
+using winrt::Windows::Storage::Streams::UnicodeEncoding;
+
+namespace Microsoft::React
+{
+WinRTWebSocketResource::WinRTWebSocketResource(Uri&& uri)
+  : m_uri{ std::move(uri) }
+  , m_socket{}
+  , m_writer{ m_socket.OutputStream() }
+  , m_closeCode{ CloseCode::None }
+{
+  m_socket.MessageReceived({ this, &WinRTWebSocketResource::OnMessageReceived });
+  m_socket.Closed({ this, &WinRTWebSocketResource::OnClosed });
+
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::Untrusted);
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::InvalidName);
+
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) })
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Stop();
+}
+
+#pragma region Private members
+
+fire_and_forget WinRTWebSocketResource::PerformConnect()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    co_await m_socket.ConnectAsync(m_uri);
+
+    if (m_connectHandler)
+    {
+      m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
```

Comment:
Why not just put the logic here in the above catch?

Reply:
Same below. Feels weird to try to concern exceptions into control flow to then only do something in the error case when e already have the catch

Reply:
If I understand correctly from the UWP implementation, which uses this exact same logic, there may be a non-throwing call to the async method that still returns a non-success `hresult`.

So, the error hast two possible sources:
1. Exception
2. `hr` values for which `SUCCEEDED` evaluates to `false`.

Reply:
We seem to only ever set the hresult in the catch block?

Reply:
Oh, my bad.
Disregard the last response.
This can and will be simplified.